File: 122_medimum_best_time_to_buy_and_sell_stock_II/main.cpp
#include "../include/leetcode.hpp"

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int i = 1, p;
        
        while(i < prices.size()) {
            p = i;
            while(i + 1 < prices.size() && prices[i + 1] >= prices[i])
                i++;
            profit += prices[i] - prices[p];
            i++;
        }

        return profit;
    }
};

File: 122_medimum_best_time_to_buy_and_sell_stock_II/main.rs
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let n = prices.len();
        let mut dp = vec![vec![0; n]; 2];

        dp[1][0] = -prices[0];
        for j in 1..n {
            dp[0][j] = dp[0][j - 1].max(dp[1][j - 1] + prices[j]);
            dp[1][j] = dp[1][j - 1].max(dp[0][j - 1] - prices[j]);
        }

        println!("{:?}", dp);
        dp[0][n - 1]
    }
}

fn main() {
    let prices = vec![7,1,5,3,6,4];
    println!("{}", Solution::max_profit(prices));
}
